### PR TITLE
Improve error message for wrong target names

### PR DESCRIPTION
**Pull request by hi-rustin**
### What does this PR try to resolve?

Fixes https://github.com/rust-lang/cargo/issues/10982

We can print all available targets for users.

### How should we test and review this PR?

- [x] unit test

### CODE REVIEW

**Comment by hi-rustin** [comment]
I'll add a unit test later.

File: src/cargo/ops/cargo_compile.rs [hunk 1]
```diff
@@ -1526,19 +1527,38 @@ fn find_named_targets<'a>(
...
+                "no {} target {} `{}`.\n",
+                target_desc,
+                if is_glob { "matches pattern" } else { "named" },
+                target_name,
+            )?;
+            writeln!(msg, "Available {} targets:", target_desc,)?;
+            for target in targets {
+                writeln!(msg, "    {}", target.name())?;
+            }
+            anyhow::bail!(msg);
```
Comment (epage) [inline src/cargo/ops/cargo_compile.rs:1560]: > I'll add a unit test later.

*(personally, I find comments on the PR diff easier to track than comments in the PR thread)*

Comment (hi-rustin) [reply src/cargo/ops/cargo_compile.rs]: Got it. Thanks! I'll add it in PR diff next time.

File: src/cargo/ops/cargo_compile.rs [hunk 2]
```diff
@@ -1526,19 +1527,41 @@ fn find_named_targets<'a>(
...
+        )?;
+        if targets.is_empty() {
+            writeln!(msg, "No {} targets available.", target_desc)?;
+            anyhow::bail!(msg);
+        }
+        writeln!(msg, "Available {} targets:", target_desc)?;
+        for target in targets {
+            writeln!(msg, "    {}", target.name())?;
+        }
+        anyhow::bail!(msg);
```
Comment (epage) [inline src/cargo/ops/cargo_compile.rs:1564]: nit: these uses early returns / fall throughs when dealing with mutually exclusive high level requirements.  I've found that doing so can make the code harder to read.  Instead highlighting this with if/else's helps draw the readers to it to better understand the structure of the code.

Someone might ask "isn't that banning early returns?".  Not quite.  The difference is in what the code is doing.  For bookkeeping, like handling an error case, you want to draw as little attention to it as possible, so early returns make sense.  When its the core logic of the function, you instead want to call attention to it.  Of course there are different situations where this doesn't always work out in an ideal way but its a starting point I've found useful for more reasdable code.

This applies both to
- `if !suggestions.is_empty()`
- `if targets.is_empty()`

File: src/cargo/ops/cargo_compile.rs [hunk 3]
```diff
@@ -1526,19 +1527,41 @@ fn find_named_targets<'a>(
...
             if is_glob { "matches pattern" } else { "named" },
             target_name,
-            suggestion
-        );
+        )?;
+        if targets.is_empty() {
+            writeln!(msg, "No {} targets available.", target_desc)?;
+            anyhow::bail!(msg);
+        }
+        writeln!(msg, "Available {} targets:", target_desc)?;
```
Comment (epage) [inline src/cargo/ops/cargo_compile.rs:1560]: nit: It looks like we are inconsistent on whether to capitalize this.  Let's lean towards not capitalizing for now

File: src/cargo/ops/cargo_compile.rs [hunk 4]
```diff
@@ -1526,19 +1527,41 @@ fn find_named_targets<'a>(
...
             target_desc,
             if is_glob { "matches pattern" } else { "named" },
             target_name,
-            suggestion
-        );
+        )?;
+        if targets.is_empty() {
+            writeln!(msg, "No {} targets available.", target_desc)?;
+            anyhow::bail!(msg);
+        }
```
Comment (epage) [inline src/cargo/ops/cargo_compile.rs:1559]: I feel like we should find a way to collapse these messages
- There is a nuance to them that can easily be lost and someone can instead think they are redundant
- The fact that they overlap quite a bit in intention makes it so the signal-to-noise ratio is lower.

**Review by epage — COMMENTED** [review]
The main "blocker" is the collapsing of messages.  The others you can handle as you wish.

And I put "blocker" in quotes because I am willing to bend on it depending on what peoples thoughts are

File: src/cargo/ops/cargo_compile.rs (hunk 2 above)
Comment (hi-rustin) [reply src/cargo/ops/cargo_compile.rs]: Make sense. Addressed.

File: src/cargo/ops/cargo_compile.rs (hunk 4 above)
Comment (hi-rustin) [reply src/cargo/ops/cargo_compile.rs]: Make sense. The latter one was deleted.

File: src/cargo/ops/cargo_compile.rs (hunk 3 above)
Comment (hi-rustin) [reply src/cargo/ops/cargo_compile.rs]: It refers to:
```log
> cargo run --bin
error: "--bin" takes one argument.
Available binaries:
    abc
    xyz
```

So I prefer to keep it.

**Comment by epage** [comment]
Thanks!

@bors r+

**Pull request merged**